backend/apps/gateway/src/websocket-gateways/yjs.gateway.ts

```diff
@@ -139,17 +164,26 @@ export class YjsGateway extends BaseWebsocketGateway {
 
         // send the persisted data to clients
         Y.applyUpdate(ydoc, Y.encodeStateAsUpdate(persistedYdoc));
-
         // store updates of the document in db
         ydoc.on('update', async (update) => {
           mdb.storeUpdate(docName, update);
         });
 
         persistedYdoc.destroy();
       },
-      writeState: async (docName) => {
+      writeState: async (docName, ydoc) => {
         // This is called when all connections to the document are closed.
 
+        // Save attempt to history service
+        const yText = ydoc.getText('monaco').toString();
+
+        await historyService
+          .createHistoryAttempt({
+            sessionId: docName,
+            questionAttempt: yText,
+          })
+          .forEach(({ histories }) => histories);
+
         // flush document on close to have the smallest possible database
         await mdb.flushDocument(docName);
       },
```

Comment:
This might be buggy for now due to #87 (we close all ws during language change, but the session has not ended). I am facing a similar issue for closing of sessions. Let me see if i can resolve this

Reply:
Addressed in #90